### PR TITLE
Add ComponentEditorPage tests and fix some bugs

### LINKED ISSUE #1356: ComponentEditorPage.SetSite(null) throws NRE should set the site to null
```cs
[Fact]
public void ComponentEditorPage_SetSite_Invoke_SetsPageSite()
{
    var control = new Control();
    var controlSite = new Mock<IComponentEditorPageSite>(MockBehavior.Strict);
    controlSite
        .Setup(s => s.GetControl())
        .Returns(control)
        .Verifiable();
    var noControlSite = new Mock<IComponentEditorPageSite>(MockBehavior.Strict);
    noControlSite
        .Setup(s => s.GetControl())
        .Returns<Control>(null)
        .Verifiable();
    var page = new SubComponentEditorPage();
    page.SetSite(controlSite.Object);
    Assert.Same(controlSite.Object, page.PageSite);
    Assert.Same(page, Assert.Single(control.Controls));
    Assert.Equal(0, page.LoadComponentCallCount);
    Assert.Equal(0, page.Loading);
    Assert.False(page.LoadRequired);

    // Set same.
    page.SetSite(controlSite.Object);
    Assert.Same(controlSite.Object, page.PageSite);
    Assert.Same(page, Assert.Single(control.Controls));
    Assert.Equal(0, page.LoadComponentCallCount);
    Assert.Equal(0, page.Loading);
    Assert.False(page.LoadRequired);

    // Set different.
    page.SetSite(noControlSite.Object);
    Assert.Same(noControlSite.Object, page.PageSite);
    Assert.Same(page, Assert.Single(control.Controls));
    Assert.Equal(0, page.LoadComponentCallCount);
    Assert.Equal(0, page.Loading);
    Assert.False(page.LoadRequired);

    // Set null.
    page.SetSite(null);
    Assert.Null(page.PageSite);
    Assert.Same(page, Assert.Single(control.Controls));
    Assert.Equal(0, page.LoadComponentCallCount);
    Assert.Equal(0, page.Loading);
    Assert.False(page.LoadRequired);
}

private class SubComponentEditorPage : ComponentEditorPage
{
    public new IComponentEditorPageSite PageSite
    {
        get => base.PageSite;
        set => base.PageSite = value;
    }
}
```

### LINKED ISSUE #1357: ComponentEditorPage.SetDirty() throws NRE if PageSite is null
Should be a nop

```cs
[Theory]
[InlineData(-1)]
[InlineData(0)]
[InlineData(1)]
public void ComponentEditorPage_SetDirty_InvokeWithoutPageSite_Nop(int loading)
{
    var page = new SubComponentEditorPage
    {
        Loading = loading
    };
    page.SetDirty();
    page.SetDirty();
}

private class SubComponentEditorPage : ComponentEditorPage
{
    public new void SetDirty() => base.SetDirty();
}
```

### PR DESCRIPTION
**Pull request by hughbe**
Fixes #1356 
Fixes #1357

### CODE REVIEW

File: src/System.Windows.Forms/src/System/Windows/Forms/Design/ComponentEditorPage.cs [hunk 1]
```diff
@@ -149,206 +91,150 @@ protected override CreateParams CreateParams
...
+        protected void EnterLoadingMode() => Loading++;
 
         /// <summary>
-        ///    Decrements the loading counter, which determines whether a page
-        ///    is in loading mode.
+        /// Decrements the loading counter, which determines whether a page is in loading mode.
         /// </summary>
-        protected void ExitLoadingMode()
-        {
-            Debug.Assert(loading > 0, "Unbalanced Enter/ExitLoadingMode calls");
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/Design/ComponentEditorPage.cs]: Please don't remove asserts

**Comment by RussKie** [comment]
Please restore the debug.assert and rebase on top of the latest master to resolve the conflict.
Thank you

**Comment by hughbe** [comment]
> Please restore the debug.assert and rebase on top of the latest master to resolve the conflict.

The tests will fail if the assert is added back.

**Comment by RussKie** [comment]
Please use `TraceListenerlessContext`

File: src/System.Windows.Forms/src/System/Windows/Forms/Design/ComponentEditorPage.cs (hunk 1 above)
Comment (hughbe) [reply src/System.Windows.Forms/src/System/Windows/Forms/Design/ComponentEditorPage.cs]: I thought about this more. I think this Debug Assert is helpful to an extent - we wouldn't want loading to be negative!

Therefore I fixed behavior in release and debug mode to clamp the value between 0 and int.MaxValue. We could throw, but I don't see this as particularily helpful

**Comment by RussKie** [comment]
Thank you

**Pull request merged**